File: backend/schemas/command_center.py
"""Command Center API response schemas."""

from __future__ import annotations

from pydantic import BaseModel, Field


class DatabaseHealth(BaseModel):
    """Database connection pool health."""

    healthy: bool
    latency_ms: float
    pool_active: int
    pool_size: int
    pool_overflow: int
    migration_head: str | None = None


class RedisHealth(BaseModel):
    """Redis connection health."""

    healthy: bool
    latency_ms: float
    memory_used_mb: float | None = None
    memory_max_mb: float | None = None
    total_keys: int | None = None


class McpHealth(BaseModel):
    """MCP subprocess health."""

    healthy: bool
    mode: str
    tool_count: int
    restarts: int
    uptime_seconds: int | None = None


class CeleryHealth(BaseModel):
    """Celery worker health."""

    workers: int | None = None
    queued: int | None = None
    beat_active: bool | None = None


class LangfuseHealth(BaseModel):
    """Langfuse tracing health."""

    connected: bool = False
    traces_today: int = 0
    spans_today: int = 0


class SystemHealthZone(BaseModel):
    """Zone 1: System health dashboard."""

    status: str  # "ok" | "degraded"
    database: DatabaseHealth
    redis: RedisHealth
    mcp: McpHealth
    celery: CeleryHealth
    langfuse: LangfuseHealth


class ApiTrafficZone(BaseModel):
    """Zone 2: HTTP API traffic metrics."""

    window_seconds: int = 300
    sample_count: int = 0
    rps_avg: float = 0
    latency_p50_ms: float | None = None
    latency_p95_ms: float | None = None
    latency_p99_ms: float | None = None
    error_rate_pct: float | None = None
    total_requests_today: int = 0
    total_errors_today: int = 0
    top_endpoints: list[dict] = Field(default_factory=list)


class TierHealth(BaseModel):
    """LLM tier health status."""

    model: str
    status: str
    failures_5m: int = 0
    successes_5m: int = 0
    cascade_count: int = 0
    latency: dict = Field(default_factory=dict)


class TokenBudgetStatus(BaseModel):
    """Per-model token budget utilization."""

    model: str
    tpm_used_pct: float = 0
    rpm_used_pct: float = 0


class LlmOperationsZone(BaseModel):
    """Zone 3: LLM operations dashboard."""

    tiers: list[TierHealth] = Field(default_factory=list)
    cost_today_usd: float = 0
    cost_yesterday_usd: float = 0
    cost_week_usd: float = 0
    cascade_rate_pct: float = 0
    token_budgets: list[TokenBudgetStatus] = Field(default_factory=list)


class PipelineLastRun(BaseModel):
    """Latest pipeline run summary."""

    started_at: str
    status: str
    total_duration_seconds: float | None = None
    tickers_succeeded: int = 0
    tickers_failed: int = 0
    tickers_total: int = 0
    step_durations: dict | None = None


class PipelineWatermarkStatus(BaseModel):
    """Pipeline watermark status."""

    pipeline: str
    last_date: str
    status: str


class PipelineZone(BaseModel):
    """Zone 4: Pipeline status dashboard."""

    last_run: PipelineLastRun | None = None
    watermarks: list[PipelineWatermarkStatus] = Field(default_factory=list)
    next_run_at: str | None = None


class CommandCenterMeta(BaseModel):
    """Assembly metadata."""

    assembly_ms: int = 0
    degraded_zones: list[str] = Field(default_factory=list)


class CommandCenterResponse(BaseModel):
    """Full command center aggregate response."""

    timestamp: str
    meta: CommandCenterMeta = Field(default_factory=CommandCenterMeta)
    system_health: SystemHealthZone | dict | None = None
    api_traffic: ApiTrafficZone | dict | None = None
    llm_operations: LlmOperationsZone | dict | None = None
    pipeline: PipelineZone | dict | None = None

File: frontend/src/types/command-center.ts
export interface DatabaseHealth {
  healthy: boolean;
  latency_ms: number;
  pool_active: number;
  pool_size: number;
  pool_overflow: number;
  migration_head: string | null;
}

export interface RedisHealth {
  healthy: boolean;
  latency_ms: number;
  memory_used_mb: number | null;
  memory_max_mb: number | null;
  total_keys: number | null;
}

export interface McpHealth {
  healthy: boolean;
  mode: string;
  tool_count: number;
  restarts: number;
  uptime_seconds: number | null;
}

export interface CeleryHealth {
  workers: number | null;
  queued: number | null;
  beat_active: boolean | null;
}

export interface LangfuseHealth {
  connected: boolean;
  traces_today: number;
  spans_today: number;
}

export interface SystemHealthZone {
  status: string;
  database: DatabaseHealth;
  redis: RedisHealth;
  mcp: McpHealth;
  celery: CeleryHealth;
  langfuse: LangfuseHealth;
}

export interface TopEndpoint {
  endpoint: string;
  count: number;
}

export interface ApiTrafficZone {
  window_seconds: number;
  sample_count: number;
  rps_avg: number;
  latency_p50_ms: number | null;
  latency_p95_ms: number | null;
  latency_p99_ms: number | null;
  error_rate_pct: number | null;
  total_requests_today: number;
  total_errors_today: number;
  top_endpoints: TopEndpoint[];
}

export interface TierLatency {
  avg_ms?: number;
  p95_ms?: number;
}

export interface TierHealth {
  model: string;
  status: string;
  failures_5m: number;
  successes_5m: number;
  cascade_count: number;
  latency: TierLatency;
}

export interface TokenBudgetStatus {
  model: string;
  tpm_used_pct: number;
  rpm_used_pct: number;
}

export interface LlmOperationsZone {
  tiers: TierHealth[];
  cost_today_usd: number;
  cost_yesterday_usd: number;
  cost_week_usd: number;
  cascade_rate_pct: number;
  token_budgets: TokenBudgetStatus[];
}

export interface PipelineLastRun {
  started_at: string;
  status: string;
  total_duration_seconds: number | null;
  tickers_succeeded: number;
  tickers_failed: number;
  tickers_total: number;
  step_durations: Record<string, number> | null;
}

export interface PipelineWatermarkStatus {
  pipeline: string;
  last_date: string;
  status: string;
}

export interface PipelineZone {
  last_run: PipelineLastRun | null;
  watermarks: PipelineWatermarkStatus[];
  next_run_at: string | null;
}

export interface CommandCenterMeta {
  assembly_ms: number;
  degraded_zones: string[];
}

export interface CommandCenterResponse {
  timestamp: string;
  meta: CommandCenterMeta;
  system_health: SystemHealthZone | null;
  api_traffic: ApiTrafficZone | null;
  llm_operations: LlmOperationsZone | null;
  pipeline: PipelineZone | null;
}
